File: company-name-matching/go-examples/append-simkeys-to-file.go
package main

import (
	"bufio"
	"encoding/csv"
	"encoding/json"
	"fmt"
	"io/ioutil"
	"net/http"
	"net/url"
	"os"
)

// Replace this with your API key from https://www.interzoid.com/manage-api-account
const API_KEY = "YOUR_API_KEY_HERE"

// Hardcoded input and output file names
const inputFileName = "sample-input-file.txt"
const outputFileName = "output.csv"

func main() {
	// Open input file
	inFile, err := os.Open(inputFileName)
	if err != nil {
		fmt.Println("Error opening input file:", err)
		return
	}
	defer inFile.Close()

	// Create output CSV file
	outFile, err := os.Create(outputFileName)
	if err != nil {
		fmt.Println("Error creating output file:", err)
		return
	}
	defer outFile.Close()

	csvWriter := csv.NewWriter(outFile)
	defer csvWriter.Flush()

	// Define struct to map JSON response
	type Response struct {
		SimKey  string `json:"SimKey"`
		Code    string `json:"Code"`
		Credits string `json:"Credits"`
	}

	scanner := bufio.NewScanner(inFile)
	lineNumber := 0

	for scanner.Scan() {
		lineNumber++
		originalValue := scanner.Text()

		// Skip completely empty lines (optional; remove this if you want to process them)
		if len(originalValue) == 0 {
			continue
		}

		// URL-encode the company name
		companyParam := url.QueryEscape(originalValue)

		// Build API URL
		apiURL := "https://api.interzoid.com/getcompanymatchadvanced?license=" + API_KEY +
			"&company=" + companyParam +
			"&algorithm=model-v4-wide"

		// Call the API
		resp, err := http.Get(apiURL)
		if err != nil {
			fmt.Printf("Error calling API for line %d (%q): %v\n", lineNumber, originalValue, err)
			// Write row with empty SimKey so CSV still lines up
			_ = csvWriter.Write([]string{originalValue, ""})
			continue
		}

		body, err := ioutil.ReadAll(resp.Body)
		resp.Body.Close()
		if err != nil {
			fmt.Printf("Error reading API response for line %d (%q): %v\n", lineNumber, originalValue, err)
			_ = csvWriter.Write([]string{originalValue, ""})
			continue
		}

		var result Response
		if err := json.Unmarshal(body, &result); err != nil {
			fmt.Printf("Error parsing JSON for line %d (%q): %v\n", lineNumber, originalValue, err)
			_ = csvWriter.Write([]string{originalValue, ""})
			continue
		}

		// Optional: check result.Code for "Success"
		if result.Code != "Success" {
			fmt.Printf("Non-success code for line %d (%q): Code=%s\n", lineNumber, originalValue, result.Code)
		}

		// Write original value and SimKey as a CSV row.
		// encoding/csv will correctly escape any commas or quotes.
		if err := csvWriter.Write([]string{originalValue, result.SimKey}); err != nil {
			fmt.Printf("Error writing CSV row for line %d (%q): %v\n", lineNumber, originalValue, err)
			continue
		}
	}

	if err := scanner.Err(); err != nil {
		fmt.Println("Error reading input file:", err)
		return
	}

	fmt.Println("Done. Results written to", outputFileName)
}

File: company-name-matching/node-examples/append-simkeys-to-file.js
// Simple Node.js example that:
// - Reads company names from a text file (one per line)
// - Calls Interzoid's Company Match Advanced API for each company
// - Writes the original value and returned SimKey to a CSV file
//
// Requirements:
// - Node.js 18+ (for built-in fetch)
// - No third-party libraries

const fs = require("fs");
const readline = require("readline");

// 1. Replace this with your API key from https://www.interzoid.com/manage-api-account
const API_KEY = "YOUR_API_KEY_HERE";

// 2. Hardcoded input and output file names
const inputFileName = "sample-input-file.txt";
const outputFileName = "output.csv";

// Helper function to escape a value for CSV:
// - Wraps in double quotes if needed
// - Escapes existing double quotes by doubling them
function csvEscape(value) {
  const str = String(value);
  if (/[",\r\n]/.test(str)) {
    return `"${str.replace(/"/g, '""')}"`;
  }
  return str;
}

async function main() {
  // Open input file as a stream
  let inputStream;
  try {
    inputStream = fs.createReadStream(inputFileName, { encoding: "utf8" });
  } catch (err) {
    console.error("Error opening input file:", err.message);
    return;
  }

  // Create output stream for CSV
  let outputStream;
  try {
    outputStream = fs.createWriteStream(outputFileName, { encoding: "utf8" });
  } catch (err) {
    console.error("Error creating output file:", err.message);
    return;
  }

  // Use readline to process the input file line by line
  const rl = readline.createInterface({
    input: inputStream,
    crlfDelay: Infinity, // Handle all common line endings
  });

  let lineNumber = 0;

  for await (const line of rl) {
    lineNumber++;
    const originalValue = line;

    // Skip completely empty lines (optional; remove this if you want to process them)
    if (originalValue.length === 0) {
      continue;
    }

    // Build the API URL with proper URL-encoding
    // URLSearchParams will encode the company value safely
    const endpoint = "https://api.interzoid.com/getcompanymatchadvanced";
    const params = new URLSearchParams({
      license: API_KEY,
      company: originalValue,
      algorithm: "model-v4-wide",
    });

    const apiURL = `${endpoint}?${params.toString()}`;

    try {
      // Call the API
      const response = await fetch(apiURL);

      if (!response.ok) {
        console.error(
          `Error calling API for line ${lineNumber} (${JSON.stringify(
            originalValue
          )}): HTTP ${response.status} ${response.statusText}`
        );
        // Write row with empty SimKey so CSV still lines up
        outputStream.write(
          `${csvEscape(originalValue)},${csvEscape("")}\n`
        );
        continue;
      }

      // Read and parse JSON response
      const data = await response.json();

      // Optional: check result code for success
      if (data.Code !== "Success") {
        console.warn(
          `Non-success code for line ${lineNumber} (${JSON.stringify(
            originalValue
          )}): Code=${data.Code}`
        );
      }

      // Write original value and SimKey as a CSV row
      // csvEscape ensures commas/quotes/newlines are handled correctly
      const simKey = data.SimKey || "";
      outputStream.write(
        `${csvEscape(originalValue)},${csvEscape(simKey)}\n`
      );
    } catch (err) {
      console.error(
        `Error calling or processing API for line ${lineNumber} (${JSON.stringify(
          originalValue
        )}):`,
        err.message
      );
      // On error, still write a row with an empty SimKey
      outputStream.write(
        `${csvEscape(originalValue)},${csvEscape("")}\n`
      );
    }
  }

  // Close the output file
  outputStream.end(() => {
    console.log("Done. Results written to", outputFileName);
  });
}

// Run the script
main().catch((err) => {
  console.error("Unexpected error:", err.message);
});
